### ISSUE TITLE
ExultStudio Prompt in background makes Studio seem unresponsive

### ISSUE THREAD

**Issue by DominusExult** [exult/exult#104]
When Exult Studio gives a warning via "ExultStudio Prompt" this window is in the background of all other Studio windows. As it wants to receive a confirmation other functions of Exult Studio don't work and makes it seem unresponsive until you notice the warning window.

Example: pick for combo and exceed the 32x32 tile range.

**Comment by Dragon-Baroque** [comment]
By the existing code, the Studio Prompt dialog should appear in front of the main window of Studio. Is this true ?

But logically the Prompt dialog should appear in from of the current editor window ( in your case, the Combo Editor ), and this means an API change : in effect the Prompt / Alert method should take the owning window as a new argument., for the Prompt instance constructor to issue a proper `gtk_window_set_transient_for` to the owning window instead of the main window.

**Comment by DominusExult** [comment]
On macOS the prompt appears behind the *Exult* window, approximately where you clicked on a shape to add to the combo.

On Windows it happens similar. The prompt appears behind the Exult window where you clicked on a shape to add that exceeds the range. Due to the taskbar you can see it blinking and notice that there is a prompt. Once you brought it to front all other prompts will appear on top of the Exult window.

So I think the prompt's placement is great (right where you clicked) but that it sometimes is behind the Exult window on Windows (and always on macOS) is annoying. But I guess it is due to Exult and Exult Studio being completely different processes that this behaves like this on macOS.

**Comment by Dragon-Baroque** [comment]
I have reproduced it on Linux.
I am appending a zip file to a small patch to `studio.cc` onto Exult master, please place yourself at the base of the source git tree, type `patch -p1 < /path/to/PromptAbove.diff` and rebuild.
Now, on Linux, when the 32 tiles overflow occurs, the Prompt window shows up _in front_ of Exult, but greyed - because it belongs to Studio, which is _not_ in front. But at least it shows up visible. About the GTK API I used there are warnings that not all window managers may obey it, so please test it on macOS ( and Windows ? ).

[PromptAbove.zip](https://github.com/exult/exult/files/6263233/PromptAbove.zip)

**Comment by DominusExult** [comment]
Doesn't change anything for macOS but it seems that is a macOS windows manager problem to begin with https://stackoverflow.com/questions/47497878/gtk-window-present-does-not-move-window-to-foreground.

I'll test on Windows now

**Comment by DominusExult** [comment]
Test on Windows: it works in bringing the dialog up front, greyed out as on linux.
*But*, this interferes oddly, so clicking okay on the dialog opens another prompt as it seems to register the click on both Exult and the dialog.

**Comment by Dragon-Baroque** [comment]
Not much I can propose at this point.
Except to look at it a different way : Since the Exult window is in front as dispenser of the objects that will belong to the Combo, shouldn't Studio answer to Exult by some negative acknowledge instead of displaying the prompt window, leaving to Exult the task of displaying the problem to the user : Changing the object frame color or Blinking it or Writing the message above the object or into one of the two black bands of the Exult window ?

**Comment by DominusExult** [comment]
maybe. Let's leave it to ponder a bit

**Comment by DominusExult** [comment]
Revisiting this:

@Dragon-Baroque your suggestion of Exult handling the error seems to be the best solution. How about a new alert gump, similar to the Yes/No gump (displayed on exiting the game and when you change the resolution in the Video Options)?
Exult Studio sends the text to Exult and it displays the alert that needs to be acknowledged with OK.
Here is a mockup (washed out because I scaled it from the original small size):
![mockup](https://github.com/exult/exult/assets/3313896/09253ed2-7f39-48cc-a5b1-625886d42e0d)
(and the gump without button and text in original size):
![alert_back](https://github.com/exult/exult/assets/3313896/c13a96d9-433a-4fdf-ae35-a688bba27ad3)

This probably needs to be a bit wider, the frame color can of course be changed.

(This alert gump could potentially also replace the yes/no confirmation in the Video Options -> we don't need to use a game's assets in the Exult Setup menu?)

My reasoning behind a full gump instead of just displaying the text is that it really makes you acknowledge the message.

What do you think, Is this feasible or overkill?
If you think this a good way forward I can start to prepare the new gump.

**Comment by Dragon-Baroque** [comment]
> My reasoning behind a full gump instead of just displaying the text is that it really makes you acknowledge the message.

I agree. Moreover, it remains part of the Exult window, instead of a separate small window like the Alert prompt of Studio.

> We don't need to use a game's assets in the Exult Setup menu ?

I fail to understand your point.

For what I saw for Select_for_combo, it seems that Studio has requested Exult to enter an Edit mode, and then Exult sends messages to Studio to define the selected objects. It does not look like these messages are answered by Studio, do you confirm ? I am just trying to think about a message protocol : not only Exult should display the warning gump and enter modal mode until it is closed, but perhaps also the failing selection should be unselected ?

**Comment by DominusExult** [comment]
> > We don't need to use a game's assets in the Exult Setup menu ?
> 
> I fail to understand your point.

Disregard please :)
When you enter the Exult menu setup screen, Exult uses assets from one of the games to display elements of the gump (checkmark and the part attached to the options gump), as well as the Yes/No gump. When I wrote the above I thought it was just the Yes/No gump but it's more and that is a way into the future issue :)

> For what I saw for Select_for_combo, it seems that Studio has requested Exult to enter an Edit mode, and then Exult sends messages to Studio to define the selected objects. It does not look like these messages are answered by Studio, do you confirm ?

I have no idea here. 

> I am just trying to think about a message protocol : not only Exult should display the warning gump and enter modal mode until it is closed, but perhaps also the failing selection should be unselected ?
Yes, IMO it should be unselected, so you can continue with the combo selection once the error is acknowledged.

**Comment by DominusExult** [comment]
Patch for Alert gump attached (patch + shape in zip).
The alert only has the checkmark to close it (besides ESC), the bool alert () is not exactly needed but and it could be done differently as shown in the VideoOptions_gump.cc change. To test toggle Highdpi in Video options.
The gump may be too big.

[alert.patch](https://github.com/exult/exult/files/12582914/alert.patch)

[alert.shp.zip](https://github.com/exult/exult/files/12582911/alert.shp.zip)
<img width="1668" alt="alertshow" src="https://github.com/exult/exult/assets/3313896/a0a1066f-0ced-41ee-8e77-ea53f7a9515c">
